src/vtok_p11/src/backend/db/mod.rs

```diff
@@ -99,13 +103,37 @@ impl Db {
                 }
             }
 
-            if key_config.cert_pem.is_some() {
-                let cert = CertInfo {
-                    id: key_config.id,
-                    label: key_config.label.clone(),
-                    pem: key_config.cert_pem.clone().ok_or(Error::GeneralError)?,
-                };
-                objects.push(Object::new_trusted_cert(cert));
+            // If the provisioned private key has an optional X509 certificate
+            // or certificate chain attached to it
+            if let Some(x509_pem) = &key_config.cert_pem {
+                let x509_chain = crypto::X509Chain::chain_from_pem(&x509_pem.as_str())
+                    .map_err(Error::PemError)?;
+
+                // If multiple certs, build and verify the chain, otherwise check if it
+                // is a self-signed certificate.
+                match x509_chain.multiple_certs() {
+                    true => x509_chain.verify_chain().map_err(Error::CryptoError)?,
+                    false => x509_chain.verify(pkey).map_err(Error::CryptoError)?,
+                }
+                for (pos, x509) in x509_chain.enumerate() {
+                    // The first certificate shares the same ID with the private key
+                    // and the rest of the certificate chain get incremental IDs and
+                    // are marked as CAs.
+                    let (cert_categ, cert_id) = match pos {
+                        0 => (crypto::CertCategory::Token, key_config.id),
+                        _ => (crypto::CertCategory::Authority, key_config.id + pos),
```

Comment:
if pos = 0, `key_config.id + pos` still works. Replace this `match` with an `if`.